nomad-base/src/merkle_sync.rs

```diff
@@ -0,0 +1,151 @@
+use color_eyre::Result;
+use ethers::types::H256;
+use nomad_core::{
+    accumulator::incremental::IncrementalMerkle, db::DbError, ChainCommunicationError,
+    RawCommittedMessage,
+};
+use std::{sync::Arc, time::Duration};
+use tokio::{sync::RwLock, task::JoinHandle, time::sleep};
+use tracing::{debug, error, info, info_span, instrument::Instrumented, Instrument};
+
+use crate::NomadDB;
+
+/// Incremental merkle tree that tracks current committed root
+#[derive(Debug, Clone, Default)]
+pub struct NomadIncrementalMerkle {
+    /// Light merkle tree
+    pub tree: IncrementalMerkle,
+    /// Last committed root of tree (last signed new_root)
+    pub last_committed_root: H256,
+}
+
+impl NomadIncrementalMerkle {
+    /// Instantiate new NomadIncrementalMerkle from tree and last committed root
+    pub fn new(tree: IncrementalMerkle, last_committed_root: H256) -> Self {
+        Self {
+            tree,
+            last_committed_root,
+        }
+    }
+
+    /// Fetch the current root of the tree
+    pub fn root(&self) -> H256 {
+        self.tree.root()
+    }
+
+    /// Fetch current size of tree
+    pub fn count(&self) -> usize {
+        self.tree.count()
+    }
+
+    /// Fetch last committed root (last signed new_root)
+    pub fn last_committed_root(&self) -> H256 {
+        self.last_committed_root
+    }
+
+    /// Ingest message to tree and update last committed root
+    pub fn ingest_message(&mut self, message: &RawCommittedMessage) {
+        self.tree.ingest(message.leaf());
+        self.last_committed_root = message.committed_root;
+    }
+}
+
+/// Self-syncing light merkle tree. Polls for new messages and updates tree.
+#[derive(Debug, Clone)]
+pub struct IncrementalMerkleSync {
+    /// Self syncing merkle tree with tracked last committed root
+    pub merkle: Arc<RwLock<NomadIncrementalMerkle>>,
+    /// DB with home name key prefix
+    pub db: NomadDB,
+}
+
+/// IncrementalMerkleSync errors
+#[derive(Debug, thiserror::Error)]
+pub enum IncrementalMerkleSyncError {
+    /// IncrementalMerkleSync receives ChainCommunicationError from chain API
+    #[error(transparent)]
+    ChainCommunicationError(#[from] ChainCommunicationError),
+    /// DB Error
+    #[error("{0}")]
+    DbError(#[from] DbError),
+}
+
+impl IncrementalMerkleSync {
```

Comment:
much of this code is duplicated from `prover_sync` yes? after #35 can we replace both of these with a `TreeSync<T: Merkle<Proof = NomadProof>>`